### ISSUE TITLE
Review invoicing merge logic

### ISSUE THREAD

**Issue by sbrossie** [killbill/killbill#663]
The merge algorithm in the invoicing system that has been implemented on top of the tree is quite difficult to follow and needs to be re-thought (and re-implemented). In particular:

* The callback [shouldInsertNode](https://github.com/killbill/killbill/blob/killbill-0.17.8/invoice/src/main/java/org/killbill/billing/invoice/tree/ItemsNodeInterval.java#L201) is hard to follow (and the first `if` does not really make sense)
* Related to previous bullet, the result of such callback will drive whether [items should stay in tree](https://github.com/killbill/killbill/blob/killbill-0.17.8/invoice/src/main/java/org/killbill/billing/invoice/tree/SubscriptionItemTree.java#L167) or be kept outside to become `resulting` items. We should review whether keeping everything in the tree would not be a better approach and why currently this does not work.

Taking a step back, we could simply revisit what is the best way to merge (compacted) existing items and proposed items (and whether the tree is really the best implementation).

Let's remember, the cases of partial repair and invoice item adjustment which makes the situation more complicated. See #654.

**Comment by sbrossie** [comment]
A good approach requires a solid design that handles all known cases and also coming up with a test plan (that covers all scenario) we can think of. Lots of tests have been written in [TestSubscriptionItemTree](https://github.com/killbill/killbill/blob/killbill-0.17.8/invoice/src/test/java/org/killbill/billing/invoice/tree/TestSubscriptionItemTree.java#L44) that could help for coming up with the plan.

**Comment by sbrossie** [comment]
Finally, slightly orthogonal to that task would be to come up with scenario where we have bad data on disk. Some examples:
* Unexpected existing item covering part of a period (smaller that period that should be invoiced)
* Unexpected existing item that spans over boundary (spans over two proposed items)

**Comment by pierre** [comment]
> The merge algorithm in the invoicing system that has been implemented on top of the tree is quite difficult to follow and needs to be re-thought (and re-implemented)

As an alternative, I would suggest only clarifying / cleaning up the code for now. A lot of work has been put into https://github.com/killbill/killbill/issues/664, which might be lost by a re-implementation.

Of course, if the code is still confusing after this effort, we should revisit it.

> * The callback [shouldInsertNode](https://github.com/killbill/killbill/blob/killbill-0.17.8/invoice/src/main/java/org/killbill/billing/invoice/tree/ItemsNodeInterval.java#L201) is hard to follow (and the first `if` does not really make sense)
> * Related to previous bullet, the result of such callback will drive whether [items should stay in tree](https://github.com/killbill/killbill/blob/killbill-0.17.8/invoice/src/main/java/org/killbill/billing/invoice/tree/SubscriptionItemTree.java#L167) or be kept outside to become `resulting` items. We should review whether keeping everything in the tree would not be a better approach and why currently this does not work.

I'll address these two points specifically.

> Finally, slightly orthogonal to that task would be to come up with scenario where we have bad data on disk.

This has been fixed by https://github.com/killbill/killbill/issues/664.

**Comment by pierre** [comment]
> We should review whether keeping everything in the tree would not be a better approach and why currently this does not work.

In merge mode, we use a shallow tree to:

1. match existing and proposed items
2. generate the repair items

It's not a fully featured tree with the ability to manage new items (nodes only contain `CANCEL` items), so keeping everything in the tree isn't currently straightforward. Additionally, it's not obvious what it would give us. For example, if the existing list contains 2 items [2016-09-08;2016-09-09] and [2016-09-09;2016-10-08], and the proposed item is [2016-09-08;2016-10-08], the current code generates 2 repairs (tree logic) and keep the proposed item as-is. If we want to keep the proposed as-is anyways, why bother adding them in the tree?

An alternative implementation would be to not use the tree at all. Solving 1. above is trivial via `walkTree`/`findNode` on the tree built from the existing items. 2. however would require writing code to manage intervals -- not sure it would be simpler than using the tree functions.
